vic/src/plugins/routing/get_basins.c

```diff
@@ -112,4 +112,93 @@ void get_basins(basin_struct *basins){
     
     free(direction);
     free(river);
+}
+
+void get_basins_file(basin_struct *basins){
+    extern domain_struct global_domain;
+    extern filenames_struct filenames;
+    
+    int    *basin_list = NULL;
+    bool    duplicate;
+    
+    size_t  d2count[2];
+    size_t  d2start[2];
+    
+    size_t  i;
+    size_t  j;
+    
+    d2start[0] = 0;
+    d2start[1] = 0;
+    d2count[0] = global_domain.n_ny;
+    d2count[1] = global_domain.n_nx;
+    
+    basins->basin_map = malloc(global_domain.ncells_active * sizeof(*basins->basin_map));
+    check_alloc_status(basins->basin_map, "Memory allocation error.");
+    basin_list = malloc(global_domain.ncells_active * sizeof(*basin_list));
+    check_alloc_status(basin_list, "Memory allocation error.");
+    
+    get_active_nc_field_int(&filenames.mpi, "basin", d2start, d2count,
+                     basins->basin_map);
+    
+    basins->Nbasin = 0;
+    for (i = 0; i < global_domain.ncells_active; i++) {
+        duplicate = false;
+        
+        for(j = 0; j < basins->Nbasin; j++){
+            if(basin_list[j] == basins->basin_map[i]){
+                duplicate = true;
+                break;
+            }
+        }
+        
+        if(duplicate){
+            continue;
+        }
+        
+        basin_list[basins->Nbasin] = basins->basin_map[i];
+        basins->Nbasin++;
+    }
+    
+    basins->Ncells = malloc(basins->Nbasin * sizeof(*basins->Ncells));
+    check_alloc_status(basins->Ncells, "Memory allocation error.");
+    basins->sorted_basins = malloc(basins->Nbasin * sizeof(*basins->sorted_basins));
+    check_alloc_status(basins->sorted_basins, "Memory allocation error.");  
+    
+    for(i=0;i<basins->Nbasin;i++){
+        basins->sorted_basins[i]=i;
+        basins->Ncells[i]=0;
+    }   
+    
+    for (i = 0; i < global_domain.ncells_active; i++) {        
+        for(j = 0; j < basins->Nbasin; j++){
+            if(basin_list[j] == basins->basin_map[i]){
+                basins->Ncells[j]++;
+            }
+        }
+    }
+    
+    // Sort basins by size
+    sizet_sort(basins->sorted_basins,basins->Ncells,basins->Nbasin,false);
+        
+    basins->catchment = malloc(basins->Nbasin * sizeof(*basins->catchment));
+    check_alloc_status(basins->catchment, "Memory allocation error.");
+    for(i=0;i<basins->Nbasin;i++){
+        basins->catchment[i] = malloc(basins->Ncells[i] * sizeof(basins->catchment[i]));
+        check_alloc_status(basins->catchment[i], "Memory allocation error.");
+    }
+    
+    for(i=0;i<basins->Nbasin;i++){
+        basins->Ncells[i]=0;
+    }   
+    
+    for (i = 0; i < global_domain.ncells_active; i++) {      
+        for(j = 0; j < basins->Nbasin; j++){
+            if(basin_list[j] == basins->basin_map[i]){       
+                basins->catchment[j][basins->Ncells[j]] = i;
+                basins->Ncells[j]++;
+            }
+        }
+    }
+    
+    free(basin_list);
 }
```

Comment:
Try to add a new line at the end of every file. This avoids sometimes compilation problems in non-linux platforms.